### PR TITLE
Add poison.rs to `mc-sgx-sync` crate

### PR DESCRIPTION
**Pull request by nick-mobilecoin**
Add poison.rs implementation which more or less copies from
[rust source](https://github.com/rust-lang/rust.git) at
[606c3907](https://github.com/rust-lang/rust/commit/606c3907251397a42e23d3e60de31be9d32525d5)

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #14** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/14" target="_blank">Graphite</a> 
    * **PR #13** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/13" target="_blank">Graphite</a> 
      * **PR #15** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/15" target="_blank">Graphite</a> 
        * **PR #22** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/22" target="_blank">Graphite</a> 
          * **PR #24** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/24" target="_blank">Graphite</a> 
            * **PR #30** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/30" target="_blank">Graphite</a> 
              * **PR #31** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/31" target="_blank">Graphite</a> 
                * **PR #32** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/32" target="_blank">Graphite</a> 
                  * **PR #33** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/33" target="_blank">Graphite</a> 
                    * **PR #34** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/34" target="_blank">Graphite</a> 
                      * **PR #35** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/35" target="_blank">Graphite</a> 
                        * **PR #36** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/36" target="_blank">Graphite</a>  👈
        * **PR #21** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/21" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/36?utm_source=stack-comment).

File: sync/src/poison.rs [hunk 1]
```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2023 The MobileCoin Foundation
```
Comment (nick-mobilecoin) [inline sync/src/poison.rs:1]: I'm up for suggestions on the copyright, not sure it's appropriate when this is mostly copied from https://github.com/rust-lang/rust/blob/master/library/std/src/sync/poison.rs

File: rust-toolchain.toml [hunk 2]
```diff
@@ -1,3 +1,3 @@
 [toolchain]
-channel = "nightly-2022-07-22"
+channel = "nightly-2023-01-04"
```
Comment (nick-mobilecoin) [inline rust-toolchain.toml:2]: Bumped the nightly to get `error_in_core`, looks like we missed it by a week from previous nightly, jumped to this version to match version at https://github.com/mobilecoinfoundation/mobilecoin/pull/3020

File: sync/src/poison.rs [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+//! [606c3907](https://github.com/rust-lang/rust/commit/606c3907251397a42e23d3e60de31be9d32525d5)
+//!
+//! Differences:
+//! - The imports were changed to work with the `mc-sgx` crates.
+//! - The stable attributes have been removed
+//! - Items that are crate only were converted from `pub` to `pub(crate)`
+//! - Examples that were not possible in an SGX enclave have been omitted
+//! - Ran `cargo fmt`
+
+#![allow(dead_code)]
```
Comment (nick-mobilecoin) [inline sync/src/poison.rs:14]: dead_code exclusion needed because this is the foundation for Mutex and RW lock

**Review by awygle — APPROVED** [review]

File: sync/src/poison.rs (hunk 1 above)
Comment (jcape) [reply sync/src/poison.rs:1]: I would put the Rust Foundation copyright either above or below ours...

**Review by jcape — APPROVED** [review]

File: rust-toolchain.toml (hunk 2 above)
Comment (nick-mobilecoin) [reply rust-toolchain.toml]: no longer relevant, change was made in #31

File: sync/src/poison.rs (hunk 1 above)
Comment (nick-mobilecoin) [reply sync/src/poison.rs:2]: Added rust foundation to the file, it's not clear who actually is the copyright holder due to, https://github.com/rust-lang/rust/blob/master/COPYRIGHT
> Copyrights in the Rust project are retained by their contributors. No copyright assignment is required to contribute to the Rust project.

**Review by varsha888 — APPROVED** [review]

**Comment by nick-mobilecoin** [comment]
⏳ This pull request is set to merge as part of a Graphite merge job
Stack job ID: HgREqLwB7JxPkCuBxNeY.
See details on [graphite.dev](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/36)

**Comment by nick-mobilecoin** [comment]
@varsha888 bumping you for review since I needed to restack after your review

**Review by varsha888 — APPROVED** [review]

**Pull request merged**